wolfcrypt/src/evp.c

```diff
@@ -6803,6 +6805,928 @@ void wolfSSL_EVP_PKEY_free(WOLFSSL_EVP_PKEY* key)
+    int indent, int bitlen, ASN1_PCTX* pctx)
+{
+    byte   buff[8] = { 0 };
+    word32 inOutIdx = 0;
+    word32 nSz;             /* size of modulus */
+    word32 eSz;             /* size of public exponent */
+    const byte*  n   = NULL;
+    const byte*  e   = NULL; /* pointer to modulus/exponent */
+    int idx;
+    int wsz;
+    word32 i;
+    word32 exponent = 0;
+    word32 outSz;
+    byte   outHex[3];
+    mp_int a;
+
+    (void)pctx;
+
+    if( mp_init(&a) != 0) {
+        return WOLFSSL_FAILURE;
+    }
+    if (indent < 0) {
+        indent = 0;
+    }
+    if (indent > EVP_PKEY_PRINT_INDENT_MAX) {
+        indent = EVP_PKEY_PRINT_INDENT_MAX;
+    }
+    /* parse key to get modulus and exponent */
+    if (wc_RsaPublicKeyDecode_ex(pkey, &inOutIdx, pkeySz, 
+                                            &n, &nSz, &e, &eSz) != 0) {
+        return WOLFSSL_FAILURE;
+    }
+
+    /* print out public key elements */
+    idx = 0;
+    XMEMSET(buff, 0, sizeof(buff));
+    Indent(out, indent);
+    if (wolfSSL_BIO_write(out, "RSA Public-Key: (",
+                          (int)XSTRLEN("RSA Public-Key: (")) <= 0) {
+        return WOLFSSL_FAILURE;
+    }
+    if (mp_set_int(&a, bitlen) != 0) {
+        return WOLFSSL_FAILURE;
+    }
+    if (mp_todecimal(&a, (char*)buff) != 0) {
+        return WOLFSSL_FAILURE;
+    }
+    wsz = (int)XSTRLEN((const char*)buff);
+
+    if (wolfSSL_BIO_write(out, buff + idx, wsz) <= 0) {
+        return WOLFSSL_FAILURE;
+    }
+    if (wolfSSL_BIO_write(out, " bit)\n", (int)XSTRLEN(" bit)\n")) <= 0) {
+        return WOLFSSL_FAILURE;
+    }
+    /* print Modulus */
+    Indent(out, indent);
+    if (wolfSSL_BIO_write(out, "Modulus:\n", 
+                                (int)XSTRLEN("Modulus:\n")) <= 0) { 
+        return WOLFSSL_FAILURE;
+    }
+
+    /* print modulus with lenading zero if exists */
+    if (*n & 0x80 && *(n-1) == 0) {
+        if (DumpElement(out, n - 1, nSz + 1, indent + 4) != WOLFSSL_SUCCESS) {
+            return WOLFSSL_FAILURE;
+        }
+    }
+    else {
+        if (DumpElement(out, n, nSz, indent + 4) != WOLFSSL_SUCCESS) {
+            return WOLFSSL_FAILURE;
+        }
+    }
+    /* print public Exponent */
+    idx = 0;
+    Indent(out, indent);
+    if (wolfSSL_BIO_write(out, "Exponent: ", 
+                                (int)XSTRLEN("Exponent: ")) <= 0) {
+        return WOLFSSL_FAILURE;
+    }
+    for (i = 0; i < eSz; i++) {
+        exponent <<= 8;
+        exponent += e[i];
+    }
+
+    XMEMSET(buff, 0, sizeof(buff));
+    if (mp_set_int(&a, exponent) != 0) {
+        return WOLFSSL_FAILURE;
+    }
+    if (mp_todecimal(&a, (char*)buff) != 0) {
+        return WOLFSSL_FAILURE;
+    }
+    wsz = (int)XSTRLEN((const char*)buff);
+
+    if (wolfSSL_BIO_write(out, buff + idx, wsz) <= 0) {
+        return WOLFSSL_FAILURE;
+    }
+    if (wolfSSL_BIO_write(out, " (0x", (int)XSTRLEN(" (0x")) <= 0) {
+        return WOLFSSL_FAILURE;
+    }
+    idx = 0;
```

Comment:
would mp_tohex() work better here?

Reply:
Yes it works. Replaced Base16_Encode() with mp_tohex().